tests/PHPStan/Analyser/data/bug-3991.php

```diff
@@ -18,8 +18,12 @@ public static function email($config = null)
 		assertType('array|stdClass|null', $config);
 		if (empty($config))
 		{
-			assertNativeType('mixed', $config);
-			assertType('array|stdClass|null', $config);
+			// the native type should be `0|0.0|''|'0'|array{}|false|null`
+			// the problem is that `empty($config)` translates to `!isset($config) || !$config`
+			// and before specified types of the left and right side are intersected they are "normalized"
+			// by removing the sureNotType from $scope->getType() which is `stdClass|array|null` here
+			assertNativeType('array{}|null', $config);
```

Comment:
regarding this - here `SpecifiedTypes::normalize()` wouldn't even be needed.. I think I can inline the normalization into `SpecifiedTypes::intersectWith()` and make it a bit smarter. I'd look at that in a dedicated PR

Reply:
I don't think this is right. Native type has to be evaluated alongside the full type. We're always interested in both at the same time. That means that TypeSpecifier and SpecifiedTypes should carry both at the same time.

Reply:
I think it's a stepping stone to implement my suggestions here: https://github.com/phpstan/phpstan-src/pull/1627#issuecomment-1216645119